crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -102,27 +115,75 @@ impl SettlementEncoder {
         let sell_token_index = self
             .token_index(order.order_creation.sell_token)
             .expect("missing sell token with price");
+        let (trade, execution) = match is_liquidity_order {
+            false => {
+                // settle normal order with the uniform clearing price
 
-        let buy_price = self
-            .clearing_prices
-            .get(&order.order_creation.buy_token)
-            .context("settlement missing buy token")?;
-        let buy_token_index = self
-            .token_index(order.order_creation.buy_token)
-            .expect("missing buy token with price");
-
-        let trade = Trade {
-            order,
-            sell_token_index,
-            buy_token_index,
-            executed_amount,
-            scaled_fee_amount,
-            is_liquidity_order,
+                let buy_price = self
+                    .clearing_prices
+                    .get(&order.order_creation.buy_token)
+                    .context("settlement missing buy token")?;
+                let buy_token_index = self
+                    .token_index(order.order_creation.buy_token)
+                    .expect("missing buy token with price");
+
+                let trade = Trade {
+                    order,
+                    sell_token_index,
+                    buy_token_index: UniformClearingPriceIndex(buy_token_index),
+                    executed_amount,
+                    scaled_fee_amount,
+                    is_liquidity_order,
+                };
+                let execution = trade
+                    .executed_amounts(*sell_price, *buy_price)
+                    .context("impossible trade execution")?;
+                (trade, execution)
+            }
+            true => {
+                // Liquidity orders are settled at their limit price. We set:
+                // buy_price =  sell_price * order.sellAmount / order.buyAmount, where sell_price is given from uniform clearing prices
+
+                // Rounding error checks:
+                // Following limit price constraint is checked in the smart contract:
+                // order.sellAmount.mul(sellPrice) >= order.buyAmount.mul(buyPrice),
+                // For each order, we get
+                // order.sellAmount.mul(sellPrice)  >= order.buyAmount.mul(buyPrice)
+                // <=> order.sellAmount.mul(sellPrice)  >= order.buyAmount.mul(sell_price * order.sellAmount / order.buyAmount)
```

Comment:
So on the right side buyAmount is cancelled out and we're left with sellAmount * sellPrice >= sellAmount * sellPrice ?
And since division buy buyAmount happens first before it is multiplied out any rounding error goes down?
If that is right we should spell it out the comment here to make it clear.